### PR TITLE
Replace openssl aes encryption/decryption/key creation implementations with our own

### PR DESCRIPTION
**Pull request by theuni**
In the spirit of dropping openssl where possible...

This uses @sipa's AES impelementation from #5885. As I understand it, he is comfortable with pulling this part in before the rest of fortuna.

The bulk of changes here are tests, the actual changes are small. Tests verify that encryption/decryption/keygen work as expected, but also that the results match openssl's exactly (including output buffers in failure cases).

The aes-cbc decryption attempts to be done in constant-time, in case it's re-used in the future in places that may be susceptible to timing attacks. It works with/without padding, so it could be re-used in different contexts.

@sipa I copied your structures for AES_Decrypt/AES_Encrypt for the sake of simplicity, so there's a good bit of duplication here now. I'm happy to re-work with templates, merged classes, separate files, etc.

Besides the unit tests, I've also done some quick sanity checks via rpc to verify that switching back and forth between implementations works as expected. Wallets encrypted with openssl still decrypt fine, and likewise the other way around.

### CODE REVIEW

**Comment by luke-jr** [comment]
Is there any reason to be embedding a copy of this code rather than using *some* library? This isn't even consensus-critical...

**Comment by laanwj** [comment]
@luke-jr Well we're ending up with our own implementation of AES either way, either through #5885 or this. 

You do have a point that we should not start importing all non-consensus critical dependencies as well. NIH syndrome is an enduring source of irritation to me. But using an external crypto library just for AES is also undesirable. Crypto code is very stable, it is unlikely to require maintenance, so doesn't add to the maintenance burden much.

File: src/wallet/crypter.cpp [hunk 1]
```diff
@@ -47,57 +73,40 @@ bool CCrypter::SetKey(const CKeyingMaterial& chNewKey, const std::vector<unsigne
...
-    EVP_CIPHER_CTX_init(&ctx);
-    if (fOk) fOk = EVP_EncryptInit_ex(&ctx, EVP_aes_256_cbc(), NULL, chKey, chIV) != 0;
-    if (fOk) fOk = EVP_EncryptUpdate(&ctx, &vchCiphertext[0], &nCLen, &vchPlaintext[0], nLen) != 0;
-    if (fOk) fOk = EVP_EncryptFinal_ex(&ctx, (&vchCiphertext[0]) + nCLen, &nFLen) != 0;
-    EVP_CIPHER_CTX_cleanup(&ctx);
-
-    if (!fOk) return false;
+    AES256CBCEncrypt enc(chKey, chIV, true);
+    nLen = enc.Encrypt(&vchPlaintext[0], vchPlaintext.size(), &vchCiphertext[0]);
+    if(nLen < vchPlaintext.size())
```
Comment (jonasschnelli) [inline src/wallet/crypter.cpp]: 'int' and 'size_type' comparison.
Gives a clan warning.

**Comment by jonasschnelli** [comment]
Tests are broken on mac osx. Could be mac only because travis did not report any issues. Wondering how this behaves on windows.

make check log: https://gist.github.com/jonasschnelli/a8aaf63a219f010dbc1f
Compile log: https://gist.github.com/jonasschnelli/8f947da1148b79950ead

I think further test does not make sense unless the unit-tests are running through.

**Comment by theuni** [comment]
@jonasschnelli That should fix your issue. It wasn't actually related to osx.

The tests turned up an unintended change in openssl behavior for version 1.0.1j, which was later reverted for 1.0.1k. It should be harmless since it only affects the output of failed decrypts which shouldn't be used anyway, but the tests here go to extra lengths to ensure that our output always matches openssl's.

Thanks for testing and reporting, that was a fun one to track down.

**Comment by jonasschnelli** [comment]
@theuni Tests running through now.
Slightly testes ACK (create encrypted wallets, exchanged between master-built-bitcoind and this PR bitcoind, dumped wallet, compared keys [tested wallet.dat exchanging in both directions]).

File: src/crypto/aes.cpp [hunk 2]
```diff
@@ -0,0 +1,1263 @@
...
+        return 0;
+
+    if (!pad && padsize != 0)
+        return 0;
+
+    memcpy(mixed, iv, AES_BLOCKSIZE);
+
+    // Write all but the last block
+    while(written + AES_BLOCKSIZE <= size)
+    {
```
Comment (sipa) [inline src/crypto/aes.cpp]: Coding style nit: braces on the same line

**Comment by jgarzik** [comment]
lightly tested (at branch point) ACK

Meta:  Should probably run all new files through the clang formatter, catch all nits ahead of time.

File: src/crypto/aes.cpp [hunk 3]
```diff
@@ -0,0 +1,1263 @@
...
+    while(written != size)
+    {
+        dec.Decrypt(out, data + written);
+        for (int i = 0; i != AES_BLOCKSIZE; i++)
+            *out++ ^= prev[i];
+        prev = data + written;
+        written += AES_BLOCKSIZE;
+    }
+
+    // When decrypting padding, attempt to run in constant-time
```
Comment (sipa) [inline src/crypto/aes.cpp]: Nice tricks for constant-time behavior.

**Comment by sipa** [comment]
Concept ACK. Code review ACK. I didn't test or review the tests.

**Comment by Diapolo** [comment]
@jgarzik Reminds me that we wanted to clang-format-all-sources... when will this be done finally?

**Comment by theuni** [comment]
@jgarzik Thanks for the good suggestion. I formatted the new aes files with clang-format-3.5, ignoring the wonky array init change for now as discussed on IRC.

Separate commits to avoid the need to re-ACK the original changes.

**Comment by jtimon** [comment]
Concept ACK. Needs rebase

**Comment by theuni** [comment]
Rebased. I'm not sure if this is going to end up being useful though, with @jonasschnelli's work towards a new wallet.

**Comment by jonasschnelli** [comment]
@theuni: i really like this. Using this would allow encryption of users wallets without adding openssl (or similar) as dependency. I don't have the experience to judge the risks, but if im right, the AES part is very stable and foreseeable. The critical part (PRNG) is untouched.

My plans is to use this for the `corewallet/crypter.cpp` stuff. If someone decides to not compile the "traditional" wallet, he could get rid of openssl (unless he uses QT bip70 support).

**Comment by jtimon** [comment]
Needs rebase. Maybe on top of #6954 ?

**Comment by sipa** [comment]
Needs rebase, indeed.

**Comment by laanwj** [comment]
Tagged for 0.13

**Comment by laanwj** [comment]
Time to start merging these.
Needs rebase, though.

**Comment by theuni** [comment]
rebased

**Comment by laanwj** [comment]
Awesome, thanks @theuni

**Comment by jonasschnelli** [comment]
Nice! Going to re-test soon.
What about adding the NIST test vectors to the test code. I have create a C test script recently with most NIST AES test vector. https://github.com/libbtc/libbtc/blob/master/test/aes_tests.c?

**Comment by theuni** [comment]
@jonasschnelli Some of the cbc test vectors are included here: https://github.com/bitcoin/bitcoin/pull/5949/files#diff-e425713e54c6cfd3f6056f549af0c5e4R368

Those come from http://csrc.nist.gov/publications/nistpubs/800-38a/sp800-38a.pdf

Happy to add the others.

**Comment by sipa** [comment]
utACK

File: src/test/crypto_tests.cpp [hunk 4]
```diff
@@ -16,6 +17,8 @@
 
 #include <boost/assign/list_of.hpp>
 #include <boost/test/unit_test.hpp>
+#include <openssl/aes.h>
```
Comment (paveljanik) [inline src/test/crypto_tests.cpp]: Do we need these openssl includes here?

**Comment by sipa** [comment]
In https://github.com/sipa/bitcoin/commits/const_aes you can find a rebased version of this PR on top of https://github.com/sipa/bitcoin/commit/b42f5fd6e256a8d1e789b010f25d4d389bdb691e which implements a constant-time AES implementation using the bitsliced approach from http://www.iacr.org/archive/ches2009/57470001/57470001.pdf. It's 15 to 25 times slower than the table-based approach used in #5885, but can still do ~2 MiB/s for short messages and ~10 MiB/s for long ones.

**Comment by laanwj** [comment]
Closing in favor of #7689

**Pull request closed without merging**